mnefun/_mnefun.py

```diff
@@ -2330,3 +2332,144 @@ def plot_reconstruction(evoked, origin=(0., 0., 0.04)):
     fig.tight_layout()
     return fig
+
+
+def plot_chpi_snr_raw(raw, win_length, n_harmonics):
+    """Compute and plot cHPI SNR from raw data
+
+    Parameters
+    ----------
+    win_length : float
+        Length of window to use for SNR estimates (seconds). A longer window
+        will naturally include more low frequency power, resulting in lower
+        SNR.
+    n_harmonics : int
+        Number of line frequency harmonics to include in the model.
+
+    Returns
+    -------
+    fig : Instance of matplotlib.figure.Figure
+        cHPI SNR as function of time, residual variance.
+
+    Notes
+    -----
+    A general linear model including cHPI and line frequencies is fit into
+    each data window. The cHPI power obtained from the model is then divided
+    by the residual variance (variance of signal unexplained by the model) to
+    obtain the SNR.
+    The SNR may decrease either due to decrease of cHPI amplitudes (e.g.
+    head moving away from the helmet), or due to increase in the residual
+    variance. In case of broadband interference that overlaps with the cHPI
+    frequencies, the resulting decreased SNR accurately reflects the true
+    situation. However, increased narrowband interference outside the cHPI
+    and line frequencies would also cause an increase in the residual variance,
+    even though it wouldn't necessarily affect estimation of the cHPI
+    amplitudes. Thus, this method is intended for a rough overview of cHPI
+    signal quality. A more accurate picture of cHPI quality (at an increased
+    computational cost) can be obtained by examining the goodness-of-fit of
+    the cHPI coil fits.
+    """
+
+    # plotting parameters
+    legend_fontsize = 9
+    title_fontsize = 12
+
+    # get some info from fiff
+    sfreq = raw.info['sfreq']
+    linefreq = raw.info['line_freq']
+    linefreqs = (np.arange(n_harmonics + 1) + 1) * linefreq
+    buflen = int(win_length * sfreq)
+    if buflen <= 0:
+        raise ValueError('Window length should be >0')
+    (cfreqs, _, _, _, _) = _get_hpi_info(raw.info)
+    print('\nNominal cHPI frequencies: %s Hz' % cfreqs)
+    print('Sampling frequency: %s Hz' % sfreq)
+    print('Using line freqs: %s Hz' % linefreqs)
+    print('Using buffers of %s samples = %s seconds\n'
+          % (buflen, buflen/sfreq))
+
+    pick_meg = pick_types(raw.info, meg=True, exclude=[])
+    pick_mag = pick_types(raw.info, meg='mag', exclude=[])
+    pick_grad = pick_types(raw.info, meg='grad', exclude=[])
+    nchan = len(pick_meg)
+
+    # create general linear model for the data
+    t = np.arange(buflen) / float(sfreq)
+    model = np.empty((len(t), 2+2*(len(linefreqs)+len(cfreqs))))
+    model[:, 0] = t
+    model[:, 1] = np.ones(t.shape)
+    # add sine and cosine term for each freq
+    allfreqs = np.concatenate([linefreqs, cfreqs])
+    ind = 2
+    for f in allfreqs:
+        model[:, ind] = np.cos(2*np.pi*f*t)
+        model[:, ind+1] = np.sin(2*np.pi*f*t)
+        ind += 2
```

Comment:
to save some lines (no need for `ind`):
```
for fi, f in enumerate(allfreqs):
    model[:, 2 * fi + 2] = ...
    model[:, 2 * fi + 3] = ...
```
Or if you want to be fancy use [broadcasting] (http://docs.scipy.org/doc/numpy/user/basics.broadcasting.html) to avoid the loop entirely:
```
model[:, 2::2] = np.cos(2 * np.pi * f[:, np.newaxis] * t)
model[:, 3::2] = np.sin(2 * np.pi * f[:, np.newaxis] * t)
```